IndexShardTests. testIndexCheckOnStartup logging assertion failure

<!--
Please fill out the following information, and ensure you have attempted
to reproduce locally
-->

**Build scan**:
https://gradle-enterprise.elastic.co/s/gr6v5hnejapom
**Repro line**:
gradlew ':server:test' --tests "org.elasticsearch.index.shard.IndexShardTests.testIndexCheckOnStartup" -Dtests.seed=CFA13989D09C1401 -Dtests.locale=ms-MY -Dtests.timezone=America/North_Dakota/Beulah -Druntime.java=11
**Reproduces locally?**:
yes, on windows
**Applicable branches**:
master, on windows
**Failure history**:
new failure after https://github.com/elastic/elasticsearch/pull/74233
**Failure excerpt**:
```
java.lang.AssertionError: expected to see failure message but did not
12:42:33     Expected: <true>
12:42:33          but: was <false>
12:42:33         at __randomizedtesting.SeedInfo.seed([CFA13989D09C1401:CE07EC20DA1AD6D9]:0)
12:42:33         at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
12:42:33         at org.elasticsearch.test.MockLogAppender$SeenEventExpectation.assertMatched(MockLogAppender.java:121)
12:42:33         at org.elasticsearch.test.MockLogAppender.assertAllExpectationsMatched(MockLogAppender.java:55)
12:42:33         at org.elasticsearch.index.shard.IndexShardTests.testIndexCheckOnStartup(IndexShardTests.java:3117)
```
but the logs contain the expected log lines..
```
12:42:33   1> [2021-06-18T05:42:15,619][WARN ][o.e.i.s.IndexShard       ] [testIndexCheckOnStartup] [index][0] performing expensive diagnostic checks during shard startup [index.shard.check_on_startup=checksum]; these checks should only be enabled temporarily, you must remove this index setting as soon as possible
12:42:33   1> [2021-06-18T05:42:15,625][WARN ][o.e.i.s.IndexShard       ] [testIndexCheckOnStartup] [index][0] check index [failure]
12:42:33   1> checksum passed: _0.cfe
12:42:33   1> checksum passed: _0.si
12:42:33   1> checksum failed: _0.cfs
12:42:33   1> org.apache.lucene.index.CorruptIndexException: checksum failed (hardware problem?) : expected=e72f63ce actual=df4db7c6 (resource=BufferedChecksumIndexInput(MockIndexInputWrapper(NIOFSIndexInput(path="C:\Users\jenkins\workspace\platform-support\91\server\build\testrun\test\temp\org.elasticsearch.index.shard.IndexShardTests_CFA13989D09C1401-001\tempDir-114\indices\_na_\0\index\_0.cfs")))
```

I wonder if this is windows related. It passes for me on macos

--- a/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java
+++ b/server/src/main/java/org/elasticsearch/index/shard/IndexShard.java
@@ -2575,8 +2575,6 @@ public class IndexShard extends AbstractIndexShardComponent implements IndicesCl
         if (Lucene.indexExists(store.directory()) == false) {
             return;
         }
-        BytesStreamOutput os = new BytesStreamOutput();
-        PrintStream out = new PrintStream(os, false, StandardCharsets.UTF_8.name());
 
         if ("checksum".equals(checkIndexOnStartup)) {
             // physical verification only: verify all checksums for the latest commit
@@ -2588,23 +2586,36 @@ public class IndexShard extends AbstractIndexShardComponent implements IndicesCl
                 logger.warn("check index [failure]", e);
                 throw e;
             }
+            final List<String> checkedFiles = new ArrayList<>(metadata.size());
             for (Map.Entry<String, StoreFileMetadata> entry : metadata.asMap().entrySet()) {
                 try {
                     Store.checkIntegrity(entry.getValue(), store.directory());
-                    out.println("checksum passed: " + entry.getKey());
-                } catch (IOException exc) {
-                    out.println("checksum failed: " + entry.getKey());
-                    exc.printStackTrace(out);
-                    corrupt = exc;
+                    if (corrupt == null) {
+                        checkedFiles.add(entry.getKey());
+                    } else {
+                        logger.info("check index [ok]: checksum check passed on [{}]", entry.getKey());
+                    }
+                } catch (IOException ioException) {
+                    for (final String checkedFile : checkedFiles) {
+                        logger.info("check index [ok]: checksum check passed on [{}]", checkedFile);
+                    }
+                    checkedFiles.clear();
+                    logger.warn(new ParameterizedMessage("check index [failure]: checksum failed on [{}]", entry.getKey()), ioException);
+                    corrupt = ioException;
                 }
             }
-            out.flush();
             if (corrupt != null) {
-                logger.warn("check index [failure]\n{}", os.bytes().utf8ToString());
                 throw corrupt;
+            }
+            if (logger.isDebugEnabled()) {
+                for (final String checkedFile : checkedFiles) {
+                    logger.debug("check index [ok]: checksum check passed on [{}]", checkedFile);
+                }
             }
         } else {
             // full checkindex
+            final BytesStreamOutput os = new BytesStreamOutput();
+            final PrintStream out = new PrintStream(os, false, StandardCharsets.UTF_8.name());
             final CheckIndex.Status status = store.checkIndex(out);
             out.flush();
             if (status.clean == false) {
@@ -2617,10 +2628,10 @@ public class IndexShard extends AbstractIndexShardComponent implements IndicesCl
                 logger.warn("{}", os.bytes().utf8ToString());
                 throw new IOException("index check failure");
             }
-        }
 
-        if (logger.isDebugEnabled()) {
-            logger.debug("check index [success]\n{}", os.bytes().utf8ToString());
+            if (logger.isDebugEnabled()) {
+                logger.debug("check index [success]\n{}", os.bytes().utf8ToString());
+            }
         }
 
         recoveryState.getVerifyIndex().checkIndexTime(Math.max(0, TimeValue.nsecToMSec(System.nanoTime() - timeNS)));

